pkg/controller/devicetainteviction/device_taint_eviction.go

```diff
@@ -0,0 +1,922 @@
+	"k8s.io/client-go/kubernetes/scheme"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/dynamic-resource-allocation/resourceclaim"
+	resourceslicetracker "k8s.io/dynamic-resource-allocation/resourceslice/tracker"
+	"k8s.io/klog/v2"
+	apipod "k8s.io/kubernetes/pkg/api/v1/pod"
+	"k8s.io/kubernetes/pkg/controller/devicetainteviction/metrics"
+	"k8s.io/kubernetes/pkg/controller/tainteviction"
+	utilpod "k8s.io/kubernetes/pkg/util/pod"
+)
+
+const (
+	// retries is the number of times that the controller tries to delete a pod
+	// that needs to be evicted.
+	retries = 5
+)
+
+// Controller listens to Taint changes of DRA devices and Toleration changes of ResourceClaims,
+// then deletes Pods which use ResourceClaims that don't tolerate a NoExecute taint.
+// Pods which have already reached a final state (aka terminated) don't need to be deleted.
+//
+// All of the logic which identifies pods which need to be evicted runs in the
+// handle* event handlers. They don't call any blocking method. All the blocking
+// calls happen in a [tainteviction.TimedWorkerQueue], using the context passed to Run.
+//
+// The [resourceslicetracker] takes care of applying taints defined in DeviceTaintRules
+// to ResourceSlices. This controller here receives modified ResourceSlices with all
+// applicable taints from that tracker and doesn't need to care about where a
+// taint came from, the DRA driver or a DeviceTaintRule.
+type Controller struct {
+	name string
+
+	// logger is the general-purpose logger to be used for background activities.
+	logger klog.Logger
+
+	// handlerLogger is specifically for logging during event handling. It may be nil
+	// if no such logging is desired.
+	eventLogger *klog.Logger
+
+	client        clientset.Interface
+	recorder      record.EventRecorder
+	podInformer   coreinformers.PodInformer
+	podLister     corelisters.PodLister
+	claimInformer resourceinformers.ResourceClaimInformer
+	sliceInformer resourceinformers.ResourceSliceInformer
+	taintInformer resourcealphainformers.DeviceTaintRuleInformer
+	classInformer resourceinformers.DeviceClassInformer
+	haveSynced    []cache.InformerSynced
+	metrics       metrics.Metrics
+
+	// evictPod ensures that the pod gets evicted at the specified time.
+	// It doesn't block.
+	evictPod func(pod tainteviction.NamespacedObject, fireAt time.Time)
+
+	// cancelEvict cancels eviction set up with evictPod earlier.
+	// Idempotent, returns false if there was nothing to cancel.
+	cancelEvict func(pod tainteviction.NamespacedObject) bool
+
+	// allocatedClaims holds all currently known allocated claims.
+	allocatedClaims map[types.NamespacedName]allocatedClaim // TODO: is value or pointer more efficient?
+
+	// pools indexes all slices by driver and pool name.
+	pools map[poolID]pool
+
+	hasSynced atomic.Int32
+}
+
+type poolID struct {
```

Comment:
This pool management stuff seems widely useful...maybe next cycle it can be factored out for use by other controllers?

Reply:
Of the top of my head I can't think of another controller which needs it exactly like this. Let's wait till we have a specific need and then factor out common code.

The namedspacedobject.go is something that I already needed in two places and IMHO should be alongside the NamespacedName type. But I didn't want to start a discussion around that now, so let's keep it a local type for now.